driver-core/src/main/com/mongodb/internal/operation/CommandOperationHelper.java

```diff
@@ -798,15 +501,55 @@ static boolean isNamespaceError(final Throwable t) {
         }
     }
 
-    @SuppressWarnings("BooleanMethodIsAlwaysInverted")
-    private static boolean shouldAttemptToRetryRead(final boolean retryReadsEnabled, final Throwable t) {
-        return retryReadsEnabled && isRetryableException(t);
+    private static boolean shouldAttemptToRetryRead(final RetryState retryState, final Throwable attemptFailure) {
+        Throwable failure = attemptFailure instanceof ResourceSupplierInternalException ? attemptFailure.getCause() : attemptFailure;
+        boolean decision = isRetryableException(failure);
+        if (!decision) {
+            logUnableToRetry(retryState.attachment(AttachmentKeys.operationName()).orElse(null), failure);
+        }
+        return decision;
     }
 
-    @SuppressWarnings("BooleanMethodIsAlwaysInverted")
-    private static boolean shouldAttemptToRetryWrite(@Nullable final BsonDocument command, final Throwable t,
-                                                     final int maxWireVersion) {
-        return shouldAttemptToRetryWrite(isRetryWritesEnabled(command), t, maxWireVersion);
+    static boolean shouldAttemptToRetryWrite(final RetryState retryState, final Throwable attemptFailure) {
+        Throwable failure = attemptFailure instanceof ResourceSupplierInternalException ? attemptFailure.getCause() : attemptFailure;
+        boolean decision = false;
+        MongoException exceptionRetryableRegardlessOfCommand = null;
+        if (failure instanceof MongoConnectionPoolClearedException) {
+            decision = true;
+            exceptionRetryableRegardlessOfCommand = (MongoException) failure;
+        }
+        BsonDocument writeCommand = retryState.attachment(AttachmentKeys.command()).orElse(null);
+        BulkWriteTracker bulkWriteTracker = retryState.attachment(AttachmentKeys.bulkWriteTracker()).orElse(null);
```

Comment:
Consider refactoring this method so that BulkWriteTracker is not depended on directly.

Reply:
Done.